### PR TITLE
"Don't fear the fsync()"

### PR DESCRIPTION
**Pull request by sourcejedi**
When updating files which are vital to boot, use `fsync()` to avoid opening any window where power loss will zero them out, or worse.

I know lots of app developers coded to the ext3 implementation.  But the only formal documentation we have, says we're broken if we omit the `fsync()`.  E.g. `man mount` (search for `auto_da_alloc`).

### CODE REVIEW

**Comment by Jesin** [comment]
Might it make sense to use `fdatasync()` instead for any of this?

**Comment by sourcejedi** [comment]
@Jesin going by the manpage, that would still avoid breaking the boot (in all cases).  However I think that would make WRITE_STRING_FILE_SYNC / fflush_sync_and_check() bad interfaces.  These aren't places where throughput is significant - I can agree I should have asserted that in the commit message.

IOW it's premature optimization.  The code is much simpler to read if we follow the most familiar advice, avoid littering the code with `/* we don't care if the modified timestamp for /etc/hostname is wrong */` and `/* this interface does not sync modified timestamps to disk */`.

The impression from the man page is there wouldn't be _any_ difference, since we're always changing the file size anyway.  That could be wrong, but it's the sort of question <del>you</del> people would end up asking, and I for one resent having to think about it if it's not important :).

File: src/basic/fileio.c [hunk 1]
```diff
@@ -71,7 +71,7 @@ int write_string_stream_ts(FILE *f, const char *line, bool enforce_newline, stru
         return fflush_and_check(f);
 }
 
-static int write_string_file_atomic(const char *fn, const char *line, bool enforce_newline) {
+static int write_string_file_atomic(const char *fn, const char *line, bool enforce_newline, bool sync) {
```
Comment (poettering) [inline src/basic/fileio.c]: glibc exports a function called sync(), we should probably avoid overloading that as a variable here locally (gcc even used to warn about that, not sure why it doesn't anymore), to avoid confusion around what "if (sync)" actually means...

I wonder if it wouldn't be easier to read to just change this call (and related ones) to just take a WriteStringFileFlags, to unify behaviour here... Functions with many bool params are so hard to call, since you never know which bool param you are actually passing there...

**Comment by poettering** [comment]
> Might it make sense to use fdatasync() instead for any of this?

iirc there's zero difference between fsync() and fdatasync() in the Linux implementation. @dvdhrm might know the details on this?

**Comment by dvdhrm** [comment]
There definitely is a difference between `fsync()` and `fdatasync()`. The code-paths taken are the same, but the file-system drivers are told whether a metadata sync shall be done (fsync()) or not (fdatasync()). But then again, metadata needed to access file-content must still be synced, which might end up syncing the entire metadata, since a file-system might not support writing them separately.

Long story short: Use fsync().

If there is some FS hacker around who can give some insights on why fdatasync() should be preferred, I'd be glad to listen. But until then, I think using fdatasync() would just look like cargo cult.

File: src/basic/fileio.c (hunk 1 above)
Comment (sourcejedi) [reply src/basic/fileio.c]: Well spotted.

bool parameters are horrible, but I think they work in this case.  It can only be called locally, and both callers use a syntax which has fairly clear intent.  I think it's clearer than having to work out which functions are responsible for which of the flags.  Maybe commenting the function as such and masking the relevant flags would make it ok, but I'm not sure how much that helps.

**Comment by poettering** [comment]
CI failure appears unrelated:

```
strip:/var/tmp/systemd-test.bBpQa3/root/usr/bin/kernel-install: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/usr/lib/kernel/install.d/90-loaderentry.install: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/usr/lib/kernel/install.d/85-initrd.install: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/usr/lib/kernel/install.d/50-depmod.install: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/usr/share/bug/udev/script: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/usr/share/bug/systemd/script: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/usr/share/initramfs-tools/scripts/init-bottom/udev: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/usr/share/initramfs-tools/scripts/init-top/udev: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/usr/share/initramfs-tools/hooks/udev: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/lib/systemd/systemd-sysv-install: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/lib/systemd/set-cpufreq: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/lib/systemd/systemd-sulogin-shell: File format not recognized
strip:/var/tmp/systemd-test.bBpQa3/root/etc/init.d/udev: File format not recognized
Makefile:4: recipe for target 'setup' failed
```
There appears to be something wrong with the Debian build scripts. @martinpitt, any idea?

Anyway, merging.

**Comment by poettering** [comment]
(btw, we generally try to merge things in "perfect" form, i.e. so that each individual commit of a PR makes logic sense, but doesn't really necessarily reflect history. This means your latter two patches should be merged actually, as the third is just a fixup for the second. I'll now merge this anyway, as the fixup isn't strictly necessary to keep things bisectable, but please keep this in mind for the next time! Thank you!)
